### PR TITLE
feat(campaign): allows to allocate shares for mainnet accounts

### LINKED ISSUE #265: campaign: allows to add mainnet accounts
Implement `MsgAddAccount` to create a new mainnet account for the campaign

A new `MainnetAccount` entry is created and `AllocatedShares` of the campaign is updated

### PR DESCRIPTION
**Pull request by Pantani**
Closes #265

### What does this PR does?

Implement `MsgAddAccount` to create a new mainnet account for the campaign

### CODE REVIEW

**Review by lubtd — COMMENTED** [review]
We actually didn't handle this in the specs but there is a flaw:
The coordinator can erase already existing shares for an account and can then overwrite shares that have been bought through vouchers by the users.

If we consider `AddMainnetAccount` as a shortcut for:
- Minting vouchers
- Sending vouchers to the account
- Redeeming the vouchers to the account

We should then increase the value of the existing shares of the account if it already exists

And maybe consider modifying the message name

File: x/campaign/keeper/msg_server_add_mainnet_account_test.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			msg: types.MsgAddMainnetAccount{
+				Coordinator: coordAddr1,
+				CampaignID:  100,
+				Address:     addr,
+				Shares:      sample.Shares(),
+			},
+			err: types.ErrCampaignNotFound,
+		},
+		{
+			name: "invalid coordinator address",
```
[the remark this answers is not in the document]
Comment (Pantani) [reply x/campaign/keeper/msg_server_add_mainnet_account_test.go:85]: yes, I updated the description to be more clear

**Comment by Pantani** [comment]
> We actually didn't handle this in the specs but there is a flaw:
> The coordinator can erase already existing shares for an account and can then overwrite shares that have been bought through vouchers by the users.
> 
> If we consider `AddMainnetAccount` as a shortcut for:
> 
> * Minting vouchers
> * Sending vouchers to the account
> * Redeeming the vouchers to the account
> 
> We should then increase the value of the existing shares of the account if it already exists
> 
> And maybe consider modifying the message name

Updated https://github.com/tendermint/spn/pull/304/commits/cbe887e805e3c617504c217459e7f35a4c7506ed 
What do you think for `AddSharesMsg` or `SetMainnetAccount` for the msg name?

File: x/campaign/keeper/msg_server_add_mainnet_account.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		return nil, sdkerrors.Wrapf(types.ErrMainnetInitialized, "%v", msg.CampaignID)
+	}
+
+	// check if the account already exists
+	account, found := k.GetMainnetAccount(ctx, campaign.Id, msg.Address)
+	if !found {
+		// if not, create the account
+		account = types.MainnetAccount{
+			CampaignID: campaign.Id,
+			Address:    msg.Address,
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_add_mainnet_account.go:43]:
```suggestion
			Address:    msg.Address,
			Shares: types.EmptyShares()
```

File: x/campaign/keeper/msg_server_add_mainnet_account_test.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+		campaignKeeper, _, campaignSrv, profileSrv, sdkCtx = setupMsgServer(t)
+		ctx                                                = sdk.WrapSDKContext(sdkCtx)
+	)
+
+	// create shares
+	allocatedShares, err := types.NewShares("91token")
+	require.NoError(t, err)
+	totalShares, err := types.NewShares("100token")
+	require.NoError(t, err)
+	share1, err := types.NewShares("10token")
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_add_mainnet_account_test.go:35]: Can we use name like `shareTooHigh`, `shareValid` for better tests readability?

File: x/campaign/keeper/msg_server_add_mainnet_account_test.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	campaign1.AllocatedShares = allocatedShares
+	campaign1.TotalShares = totalShares
+
+	campaignKeeper.SetMainnetAccount(sdkCtx, sample.MainnetAccount(campaign1.Id, addr2))
+	res2, err := profileSrv.CreateCoordinator(ctx, &profiletypes.MsgCreateCoordinator{
+		Address:     coordAddr2,
+		Description: sample.CoordinatorDescription(),
+	})
+	require.NoError(t, err)
+	campaign2.CoordinatorID = res2.CoordinatorId
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_add_mainnet_account_test.go:57]: Same, can we use name like `campaignMainnetInitialized`

File: x/campaign/keeper/msg_server_add_mainnet_account_test.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+				CampaignID:  campaign1.Id,
+				Address:     addr2,
+				Shares:      share1,
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			var (
+				accountExists bool
+				tmpAccount    types.MainnetAccount
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_add_mainnet_account_test.go:151]:
```suggestion
				previousAccount    types.MainnetAccount
```

File: x/campaign/keeper/msg_server_add_mainnet_account_test.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+				tmpCampaign   types.Campaign
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_add_mainnet_account_test.go:152]:
```suggestion
				previousCampaign   types.Campaign
```

File: x/campaign/types/errors.go [hunk 7]
```diff
@@ -10,4 +10,8 @@ import (
 var (
 	ErrInvalidCampaignName = sdkerrors.Register(ModuleName, 1, "invalid campaign name")
 	ErrInvalidTotalSupply  = sdkerrors.Register(ModuleName, 2, "invalid total supply")
+	ErrCampaignNotFound    = sdkerrors.Register(ModuleName, 3, "campaign not found")
+	ErrMainnetInitialized  = sdkerrors.Register(ModuleName, 4, "mainnet initialized")
+	ErrInvalidAccountShare = sdkerrors.Register(ModuleName, 5, "invalid account share")
+	ErrTotalShareLimit     = sdkerrors.Register(ModuleName, 6, "allocated share greater the total share")
```
Comment (lubtd) [inline x/campaign/types/errors.go:16]:
```suggestion
	ErrTotalSharesLimit     = sdkerrors.Register(ModuleName, 6, "allocated shares greater than total shares")
```

File: x/campaign/types/errors.go [hunk 8]
```diff
@@ -10,4 +10,8 @@ import (
 var (
 	ErrInvalidCampaignName = sdkerrors.Register(ModuleName, 1, "invalid campaign name")
 	ErrInvalidTotalSupply  = sdkerrors.Register(ModuleName, 2, "invalid total supply")
+	ErrCampaignNotFound    = sdkerrors.Register(ModuleName, 3, "campaign not found")
+	ErrMainnetInitialized  = sdkerrors.Register(ModuleName, 4, "mainnet initialized")
+	ErrInvalidAccountShare = sdkerrors.Register(ModuleName, 5, "invalid account share")
```
Comment (lubtd) [inline x/campaign/types/errors.go:15]:
```suggestion
	ErrInvalidAccountShares = sdkerrors.Register(ModuleName, 5, "invalid account shares")
```

File: x/campaign/types/shares.go [hunk 9]
```diff
@@ -54,6 +54,11 @@ func CheckShares(shares Shares) error {
 	return nil
 }
 
+// IsEqualShares returns true if the two sets of Shares have the same value
+func IsEqualShares(shares, newShares Shares) bool {
```
Comment (lubtd) [inline x/campaign/types/shares.go:58]: We should add a test for this method

**Comment by lubtd** [comment]
So as discussed in the `SPN` meeting, I think it would make sense to rename:
`MsgAddMainnetAccount` -> `MsgAddShares`
`MsgAddVestingAccount` -> `MsgAddVestingOptions`

Unlike in the `launch` module, we don't have either a genesis account or a vesting account but we can have both entry in the store. The final account to be defined in the genesis is interpreted from those two eventual entry.

So the coordinator doesn't add a new mainnet account but add shares to an account that eventually already exist.

The same for `MsgAddVestingAccount`, the coordinator doesn't obligatory add a new vesting account but associate vesting options to an account that eventually already exist.

Does this seem right for you?

**Comment by lubtd** [comment]
Otherwise, the CLI commands works very well for me, great job👍

**Comment by Pantani** [comment]
> So as discussed in the `SPN` meeting, I think it would make sense to rename the messages:
> `MsgAddMainnetAccount` -> `MsgAddShares`
> `MsgAddVestingAccount` -> `MsgAddVestingOptions`
> 
> Unlike in the `launch` module, we don't have either a genesis account or a vesting account but we can have both entry in the store. The final account to be defined in the genesis is interpreted from those two eventual entry.
> 
> So the coordinator doesn't add a new mainnet account but add shares to an account that eventually already exist.
> 
> The same for `MsgAddVestingAccount`, the coordinator doesn't obligatory add a new vesting account but associate vesting options to an account that eventually already exist.
> 
> Does this seem right for you?

make total sense for me

File: x/campaign/client/cli/tx_add_shares.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/cosmos/cosmos-sdk/client/tx"
+	"github.com/tendermint/spn/x/campaign/types"
+)
+
+var _ = strconv.Itoa(0)
+
+func CmdAddShares() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "add-shares [campaign-id] [address] [shares]",
+		Short: "Add a shares for an account",
```
Comment (lubtd) [inline x/campaign/client/cli/tx_add_shares.go:20]:
```suggestion
		Short: "Add a shares to an account",
```

**Review by lubtd — DISMISSED** [review]
👍

**Review by lubtd — DISMISSED** [review]

**Review by ilgooz — APPROVED** [review]

**Comment by lubtd** [comment]
Need to look at the failing test. We can merge it afterward :)

**Pull request merged**